### ISSUE TITLE
Automatically disable non-requested providers if --enable-direct specified

### ISSUE THREAD

**Issue by sungeunchoi** [ofiwg/libfabric#1985]
Generic providers like sockets are automatically detected during the configure step, so if you use --enable-direct for another provider, you get this warning (followed by a failure):

```
configure: WARNING: Only one provider can be chosen when using --enable-direct
```

Obviously you can explicitly disable the other providers (currently, socket, udp, rxm, verbs), but it would be nice to automatically disable any automatically detected provider that is _not_ the one requested by --enable-direct.

I'm not a configury expert, so I'm not sure how easy this is to do.

**Comment by shefty** [comment]
There was an error in the sockets provider configuration where it was always enabled.  I believe that has been fixed, so that it is not enabled if enable-direct has been selected.

Are you seeing that other providers were enabled, or just the sockets provider?

**Comment by sungeunchoi** [comment]
It was the rxm provider.  I did not check with others since our testing explicitly disables them.

**Comment by shefty** [comment]
Okay, the sockets provider was always including files as part of the build.

I don't see anywhere in the configure scripts where enable-direct would prevent the other providers from being built.  I agree that that should not be the case.  Everything else needs to be disabled.

@jsquyres might know the best option to handle this.

**Comment by sungeunchoi** [comment]
@jithinjosepkl I don't think anyone's worked on this, but FYI rxd provider also should be disabled when fabric direct is enabled for another provider.

Do you or @hppritcha (or anyone else) have any thoughts on the best way to do this?

**Comment by jsquyres** [comment]
Hmm.  I'm not so sure.

I say this because I'd ultimately love to see rxd (and others like it?) be able to be used under existing providers.  E.g., have usnic be able to use rxd under the covers on hardware that doesn't natively support RDM.

**Comment by sungeunchoi** [comment]
@jsquyres So you mean even in the case of fabric direct?  Does that mean explicitly enabling the provider?  I admit I don't quite understand how this would work.

**Comment by jsquyres** [comment]
Yes, I think so (although I admit I haven't thought this through fully).

E.g., if usnic is built in a direct mode, I'd still want it to be able to utilize rxd under the covers.

Note that this isn't exactly how rxd is today -- rxd is meant to be used as a provider in itself.  But I'd like to see rxd be able to be used *under* a provider.  Specifically: I want my customers to always write to use the "usnic" provider.  It doesn't matter to the application whether the usnic provider chooses to do functionality via emulation (e.g., via an underlying provider such as rxd) or via hardware offload (on platforms capable of it).  The application doesn't know/care -- it just always uses the usnic provider.

**Comment by sungeunchoi** [comment]
That makes sense, but it seems like separate issue (maybe there's one open already?) that possibly should be resolved before resolving this one.

**Comment by jithinjosepkl** [comment]
Yeah, currently all other providers need to be disabled explicitly when configured as fabric-direct. It should be done automatically (had updated travis configuration for this). But I guess we need to modify the configure files to get this disabled automatically. I can take a look at this one.

@jsquyres - The naming of the utility providers is not finalized :), may be we can discuss this in one of the libfabric calls and take a decision.

**Comment by hppritcha** [comment]
closed via #2692
